File: JavaScript/Logger/Logger.js
/**
 * Core Logger implementation with LogLevel enum and main logging functionality.
 */

import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

/**
 * Log levels in order of severity.
 */
export const LogLevel = {
    DEBUG: 0,
    INFO: 1,
    WARNING: 2,
    ERROR: 3,
    CRITICAL: 4
};

/**
 * Get the string representation of a log level.
 */
export function getLogLevelName(level) {
    return Object.keys(LogLevel).find(key => LogLevel[key] === level) || 'UNKNOWN';
}

/**
 * Represents a single log record with all relevant information.
 */
export class LogRecord {
    constructor(name, level, message, timestamp, extra = {}) {
        this.name = name;
        this.level = level;
        this.message = message;
        this.timestamp = timestamp;
        this.extra = extra;
    }
}

/**
 * Main Logger class that handles log messages and distributes them to handlers.
 */
export class Logger {
    constructor(name = "default") {
        this.name = name;
        this.level = LogLevel.INFO;
        this.handlers = [];
        this.propagate = true;
        this.parent = null;
        this.children = new Map();
    }

    /**
     * Set the minimum log level for this logger.
     */
    setLevel(level) {
        this.level = level;
    }

    /**
     * Add a handler to this logger.
     */
    addHandler(handler) {
        this.handlers.push(handler);
    }

    /**
     * Remove a handler from this logger.
     */
    removeHandler(handler) {
        const index = this.handlers.indexOf(handler);
        if (index > -1) {
            this.handlers.splice(index, 1);
        }
    }

    /**
     * Remove all handlers from this logger.
     */
    clearHandlers() {
        this.handlers = [];
    }

    /**
     * Get or create a child logger.
     */
    getChild(name) {
        if (!this.children.has(name)) {
            const child = new Logger(`${this.name}.${name}`);
            child.parent = this;
            child.level = this.level;
            child.handlers = [...this.handlers];
            this.children.set(name, child);
        }
        return this.children.get(name);
    }

    /**
     * Internal logging method.
     */
    _log(level, message, ...args) {
        if (level < this.level) {
            return;
        }

        // Format message with args
        let formattedMessage = message;
        if (args.length > 0) {
            formattedMessage = message.replace(/%s/g, () => args.shift());
        }

        // Create log record
        const record = new LogRecord(
            this.name,
            level,
            formattedMessage,
            new Date(),
            args.length > 0 ? args[0] : {}
        );

        // Send to handlers
        for (const handler of this.handlers) {
            try {
                handler.emit(record);
            } catch (error) {
                // Prevent infinite recursion if handler fails
                console.error(`Handler error: ${error.message}`);
            }
        }

        // Propagate to parent if enabled
        if (this.propagate && this.parent) {
            this.parent._log(level, message, ...args);
        }
    }

    /**
     * Log a debug message.
     */
    debug(message, ...args) {
        this._log(LogLevel.DEBUG, message, ...args);
    }

    /**
     * Log an info message.
     */
    info(message, ...args) {
        this._log(LogLevel.INFO, message, ...args);
    }

    /**
     * Log a warning message.
     */
    warning(message, ...args) {
        this._log(LogLevel.WARNING, message, ...args);
    }

    /**
     * Log an error message.
     */
    error(message, ...args) {
        this._log(LogLevel.ERROR, message, ...args);
    }

    /**
     * Log a critical message.
     */
    critical(message, ...args) {
        this._log(LogLevel.CRITICAL, message, ...args);
    }

    /**
     * Log an exception with traceback.
     */
    exception(message, error, ...args) {
        const extra = args.length > 0 ? args[0] : {};
        extra.exc_info = true;
        extra.exc_text = error ? error.stack : new Error().stack;
        this._log(LogLevel.ERROR, message, extra);
    }
} 

File: Python/Logger/logger.py
"""
Core Logger implementation with LogLevel enum and main logging functionality.
"""

import sys
from datetime import datetime
from enum import Enum
from typing import List, Optional, Any, Dict, TYPE_CHECKING
import traceback

if TYPE_CHECKING:
    from .handlers import Handler


class LogLevel(Enum):
    """Log levels in order of severity."""
    DEBUG = 0
    INFO = 1
    WARNING = 2
    ERROR = 3
    CRITICAL = 4

    def __str__(self):
        return self.name

    def __lt__(self, other):
        if isinstance(other, LogLevel):
            return self.value < other.value
        return NotImplemented


class Logger:
    """Main Logger class that handles log messages and distributes them to handlers."""
    
    def __init__(self, name: str = "default"):
        self.name = name
        self.level = LogLevel.INFO
        self.handlers: List['Handler'] = []
        self.propagate = True
        self.parent: Optional['Logger'] = None
        self.children: Dict[str, 'Logger'] = {}
        
    def set_level(self, level: LogLevel):
        """Set the minimum log level for this logger."""
        self.level = level
        
    def add_handler(self, handler: 'Handler'):
        """Add a handler to this logger."""
        self.handlers.append(handler)
        
    def remove_handler(self, handler: 'Handler'):
        """Remove a handler from this logger."""
        if handler in self.handlers:
            self.handlers.remove(handler)
            
    def clear_handlers(self):
        """Remove all handlers from this logger."""
        self.handlers.clear()
        
    def get_child(self, name: str) -> 'Logger':
        """Get or create a child logger."""
        if name not in self.children:
            child = Logger(f"{self.name}.{name}")
            child.parent = self
            child.level = self.level
            child.handlers = self.handlers.copy()
            self.children[name] = child
        return self.children[name]
        
    def _log(self, level: LogLevel, message: str, *args, **kwargs):
        """Internal logging method."""
        if level.value < self.level.value:
            return
            
        # Format message with args
        if args:
            message = message % args
            
        # Get the calling script name from the stack trace
        calling_script = self._get_calling_script()
            
        # Create log record
        record = LogRecord(
            name=self.name,
            level=level,
            message=message,
            timestamp=datetime.now(),
            calling_script=calling_script,
            **kwargs
        )
        
        # Send to handlers
        for handler in self.handlers:
            try:
                handler.emit(record)
            except Exception as e:
                # Prevent infinite recursion if handler fails
                sys.stderr.write(f"Handler error: {e}\n")
                
        # Propagate to parent if enabled
        if self.propagate and self.parent:
            self.parent._log(level, message, *args, **kwargs)
            
    def _get_calling_script(self) -> str:
        """Get the name of the script that called the logger, without .py extension."""
        import inspect
        import os
        
        # Get the current stack
        stack = inspect.stack()
        
        # Only skip logger implementation files
        skip_files = [
            'logger.py', 'formatters.py', 'handlers.py', 'registry.py', 'config.py'
        ]
        for frame_info in stack:
            filename = frame_info.filename
            base = os.path.basename(filename)
            if base not in skip_files and '__pycache__' not in filename:
                # Extract just the filename without path and strip .py
                script_name = base
                if script_name.endswith('.py'):
                    script_name = script_name[:-3]
                return script_name
        # Fallback to the logger name if we can't find the calling script
        return self.name
            
    def debug(self, message: str, *args, **kwargs):
        """Log a debug message."""
        self._log(LogLevel.DEBUG, message, *args, **kwargs)
        
    def info(self, message: str, *args, **kwargs):
        """Log an info message."""
        self._log(LogLevel.INFO, message, *args, **kwargs)
        
    def warning(self, message: str, *args, **kwargs):
        """Log a warning message."""
        self._log(LogLevel.WARNING, message, *args, **kwargs)
        
    def error(self, message: str, *args, **kwargs):
        """Log an error message."""
        self._log(LogLevel.ERROR, message, *args, **kwargs)
        
    def critical(self, message: str, *args, **kwargs):
        """Log a critical message."""
        self._log(LogLevel.CRITICAL, message, *args, **kwargs)
        
    def exception(self, message: str, *args, **kwargs):
        """Log an exception with traceback."""
        kwargs['exc_info'] = True
        self._log(LogLevel.ERROR, message, *args, **kwargs)


class LogRecord:
    """Represents a single log record with all relevant information."""
    
    def __init__(self, name: str, level: LogLevel, message: str, timestamp: datetime, **kwargs):
        self.name = name
        self.level = level
        self.message = message
        self.timestamp = timestamp
        self.calling_script = kwargs.get('calling_script', name)
        self.exc_info = kwargs.get('exc_info', False)
        self.extra = {k: v for k, v in kwargs.items() if k not in ['exc_info', 'calling_script']}
        
        # Add exception info if present
        if self.exc_info:
            self.exc_text = ''.join(traceback.format_exc())
        else:
            self.exc_text = None 